politeiawww/user/mysql/mysql.go

```diff
@@ -0,0 +1,1028 @@
+// Copyright (c) 2021 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package mysql
+
+import (
+	"bytes"
+	"context"
+	"database/sql"
+	"encoding/binary"
+	"encoding/hex"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/decred/politeia/politeiawww/user"
+	"github.com/decred/politeia/util"
+	"github.com/google/uuid"
+	"github.com/marcopeereboom/sbox"
+
+	// MySQL driver.
+	_ "github.com/go-sql-driver/mysql"
+)
+
+const (
+	// Database options
+	connTimeout     = 1 * time.Minute
+	connMaxLifetime = 1 * time.Minute
+	maxOpenConns    = 0 // 0 is unlimited
+	maxIdleConns    = 100
+
+	// Database user (read/write access)
+	userPoliteiawww = "politeiawww"
+
+	databaseID = "users"
+
+	// Database table names.
+	tableNameKeyValue   = "key_value"
+	tableNameUsers      = "users"
+	tableNameIdentities = "identities"
+	tableNameSessions   = "sessions"
+
+	// Key-value store keys.
+	keyVersion             = "version"
+	keyPaywallAddressIndex = "paywalladdressindex"
+)
+
+// tableKeyValue defines the key-value table.
+const tableKeyValue = `
+  k VARCHAR(255) NOT NULL PRIMARY KEY,
+  v LONGBLOB NOT NULL
+`
+
+// tableUsers defines the users table.
+const tableUsers = `
+  ID VARCHAR(36) NOT NULL PRIMARY KEY,
+  username VARCHAR(64) NOT NULL,
+  uBlob BLOB NOT NULL,
+  createdAt INT(11) NOT NULL,
+  updatedAt INT(11),
+  UNIQUE (username)
+`
+
+// tableIdentities defines the identities table.
+const tableIdentities = `
+  publicKey CHAR(68) NOT NULL PRIMARY KEY,
```

Comment:
Why is this 68 chars and not 64 chars?

Reply:
https://crypto.stackexchange.com/questions/87715/what-is-the-public-key-length-of-rsa-and-ed25519

Reply:
> A0. That link says ed25519 publickey is 68 chars, not 64, and is correct;

Reply:
They're encoding in base64.  We're encoding in hex.  Instead of looking on stackexchange for stuff like this and getting wrong answers because they're situation is slightly different, try to answer your own questions.  In this case, the better way to do it would be to just print a public key to the console and see how many characters it is.

Reply:
oh, i saw the array length, but still wanted to check online, my bad.

Reply:
but i know the answer is in the code for such stuff 💯